proposals/0271-swiftpm-binary-dependencies.md

```diff
@@ -0,0 +1,417 @@
+            ]
+        )
+     ]
+)
+```
+
+
+Secondly we propose to add a new configuration point to the package description that allows packages to opt-out of binary dependencies. This will enforce that nothing in its transitive dependencies brings a binary dependency with it.
+
+```swift
+let package = Package(
+    name: "SomeOtherPackage",
+    disallowsBinaryDependencies: true,
+    products: [
+        .library(name: "SomeOtherPackage", targets: ["SomeOtherPackageLib"])
+    ],
+    targets: [
+        .target(name: "SomeOtherPackageLib")
+     ]
+)
+```
+
+Packages are allowed to contain a mix of binary and source targets. This is
+useful when, for example, providing a pre-built or closed source C library
+alongside an open source set of Swift bindings for the library.
+
+When a package is built that depends upon any product with a binary target, the
+package manager will search the `artifacts` declaration list to find an artifact
+which matches the current build conditions. This list
+will be searched in order, and the first matching artifact will be used. It is
+the job of the package author/published to provide an appropriate set of
+artifacts for the use cases the package wishes to support. This use case will be limited to Apple platforms in the beginning. In the future, we can add support for other platforms. A potential approach is outlined in the future directions section.
+
+## Detailed design
+
+The design consists of the following key points:
+* New `PackageDescription` API for defining a binary target.
+* New parameter on a package declaration level to opt-out of binary dependencies.
+* New requirements for the `Package.resolved` file when using binary packages.
+* A new mechanism for downloading binary target artifacts.
+
+Terminology:
+
+* Technically, a *target* is binary or not. However, we anticipate that often a
+  single package will consist of either exclusively source or binary targets. We
+  will use the term *binary package* to refer to any package which contains at
+  least one binary product. Similarly, a *binary product* is one which contains
+  at least one binary target.
+
+Our design attempts to optimize for the following goals:
+
+* Ease of use for clients
+* Easy of implementation in existing SwiftPM
+* Ease of maintenance in the face of an evolving SwiftPM
+* Understandable composition with current and upcoming SwiftPM features
+* Support existing well-known occurrences of binary artifacts in the existing
+  (often iOS focused) target developer market.
+
+while keeping the following as non-goals:
+
+* Ease of production of binary packages
+* Simplicity of binary artifact distribution mechanism
+* Widespread use of binary packages
+
+## New `PackageDescription` API
+
+### BinaryTarget
+Since, a binary target is different compared to a source only target, we propose to introduce a new struct `Artifact`. This struct defines a targets associated artifacts.
```

Comment:
```suggestion
Since a binary target is different compared to a source only target, we propose to introduce a new struct `Artifact`. This struct defines a target's associated artifacts.
```